CIP-0111? | Web-Wallet Bridge - Wallet Transaction Caching (#733)
cardano-foundation/CIPs: cip-smart-tx-signing/README.md

---

> CIP: ?

Reviewer: ```suggestion
CIP: 111
```

Title: Wallet Transaction Caching
Status: Proposed
Category: Wallets
Authors:
    - Josh Marchand <josh@securitybot.info>
Implementors: []
Discussions:
 - https://github.com/cardano-foundation/CIPs/pull/733
Created: 2023-12-22
License: CC-BY-4.0
---

## Abstract

> This CIP extends [CIP-0030 Cardano dApp-Wallet Web Bridge](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030/) and [CIP-0103? | Web-Wallet Bridge - Bulk Transaction Signing](https://github.com/cardano-foundation/CIPs/pull/587) to allow wallets to provide a complete and accurate summary of transactions that spend outputs that don't exist on chain yet when possible, and clearly communicate to the user risks when they cannot.

Author: I think I agree in general, I was just trying to follow the standards for "adding on" to the existing CIPs. That said, I think that it makes more sense to not explicitly extend any one CIP, especially as this CIP should be used along side ANY CIP that involves a wallet prompting users for a signature.

## Motivation: why is this CIP necessary?

> Transaction chaining, building transactions that spend outputs of transactions that aren't settled on chain, allows dApps to quickly do several on-chain transactions instead of waiting for each indvidual transaction to settle before building and submitting the next one. This is a powerful feature that has allowed dApps to provide seamless experiences to users. However, due to the tranasaction data structure, wallets are unable to provide a complete and accurate summary of these chained transactions when prompting users for a signature.

Reviewer 2: ```suggestion
Transaction chaining, building transactions that spend outputs of transactions that aren't settled on chain, allows dApps to quickly do several on-chain transactions instead of waiting for each individual transaction to settle before building and submitting the next one. This is a powerful feature that has allowed dApps to provide seamless experiences to users. However, due to the transactions data structure, wallets are unable to provide a complete and accurate summary of these chained transactions when prompting users for a signature.
```

spelling

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

### Defintions

**Wallet**: A wallet is any software that manages a user's private keys and provides basic cryptographic operations such as signing data, for the purpose of interacting with the Cardano blockchain.

---

A wallet MUST maintain a cache of signed transactions that have not yet settled on chain. This cache MAY have a maximum size, and if the cache is full, the wallet MAY remove transactions from the cache, beginning with the oldest transaction to store a newer transaction. The wallet MAY remove cached transactions once they are seen on chain.

> If a user signs a transaction through the [CIP-30 signTx](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0030/README.md#apisigntxtx-cbortransaction-partialsign-bool--false-promisecbortransaction_witness_set) endpoint, the [CIP-103? signTxs](https://github.com/cardano-foundation/CIPs/blob/508ea0557bcd17d793da90312789165dcef8a4db/CIP-0103/README.md#apicip103signtxstxs-transactionsignaturerequest-promisecbortransaction_witness_set) endpoint, or any other CIP-30 extension that allows signing transactions, the wallet MUST cache the signed transaction with a TTL (time-to-live) greater than or equal to the [TTL of the transaction](https://github.com/IntersectMBO/cardano-ledger/blob/master/eras/babbage/impl/cddl-files/babbage.cddl#L57C8-L57C9). The wallet MUST NOT remove the transaction from the cache until the TTL has expired, the transaction has been seen on chain, or the cache is full and the wallet needs to remove transactions to make room for newer transactions.

Reviewer 3: I don't think cache size is a concern here. The cache will only contain utxos owned by the wallet, so even if we consider ttl, it is safe to assume that the size will not grow beyond manageable values, as each utxo creation will incur costs. We already implicitly allow the wallet size to grow indefinitely (because the number of utxos in a multi-address wallet is unlimited).

If we allow valid implementations to drop cache data due to size limits, then a dApp would not be able to rely on presence of any data in the cache. And an implementation with SIZE=0 will be a valid one.

Author: Sorry I missed this somehow. I think that it's really important to define this behavior in a standard. Lots of the current CIPs have undefined behaviors in edge cases and it can cause real issues across the ecosystem.

I want this standard to be comprehensive, regardless of device or wallet software. In the (rare, as you correctly point out) case that a cache has to be purged, it is important the wallet purges transactions in an appropriate manner (ascending chronologically). 

It's an excellent point that a wallet could correctly implement this CIP with a cache size of zero bytes. I think it makes sense to add a "minimum cache size" that is `n * maxTxSize`.  I would love to hear others' thoughts on what the value of `n` should be, since it will necessarily add (small) overhead to the wallet.

Reviewer 3: > n * maxTxSize

Why do we need to cache whole transactions? I think it could work with caching just UTxOs, and only those that belong to one of the user's addresses.

Author: While for the specific problem in the CIP is solved with just the UTxOs, I felt it may be useful to have the complete transaction. For example, caching the inputs as well would allow a wallet to refuse a signature on a transaction which consumes inputs which are already consumed in a cached transaction. Perhaps that is out of the scope of this CIP in particular though.

When displaying a transaction summary to the user, the wallet MUST search for the outputs referenced by inputs in the following order:

1. The latest known on-chain UTxO set
2. The cache of signed transactions
3. The set of transactions being signed

> If a transaction contains an input that references an output that cannot be found, the wallet MUST display a warning to the user that the transaction includes unidentifiable inputs. The style and wording of this warning is left up to the wallet. The wallet MAY prevent the user from signing a transaction with unidentified inputs, but this may cause issues when interacting with some dApps.

Reviewer 4: > The wallet MAY prevent the user from signing a transaction with unidentified inputs, but this may cause issues when interacting with some dApps

What potential issues are there?

Author: Many dApps use chaining with UTxOs that are owned by a smart contract or another address that the user's wallet is not aware of. If a wallet were to prevent signing anytime there was a UTxO it wasn't aware of, it would prevent interactions with many dApps.

Reviewer 4: Understood. How about being more assertive, given you're already saying "some dApps":
```suggestion
If a transaction contains an input that references an output that cannot be found, the wallet MUST display a warning to the user that the transaction includes unidentifiable inputs. The style and wording of this warning is left up to the wallet. The wallet MAY prevent the user from signing a transaction with unidentified inputs, but this will cause issues when interacting with some dApps.
```

## Rationale: how does this CIP achieve its goals?

> 

Reviewer 2: I think it is worth discussing here the alternatives to this approach, such as stronger warnings within wallets OR not letting wallets sign unknown transactions.

---
CIP: ?
Title: Wallet Transaction Caching
Status: Active
Category: Wallets
Authors:
    - Josh Marchand <josh@securitybot.info>
Implementors: N/A
Discussions:
 - N/A
Created: 2023-12-22
License: CC-BY-4.0
---

> # CIP-? Wallet Transaction Caching
> 

Reviewer: ```suggestion
```
CIP-0001 says "no H1 title` generally since this is constructed from the header in other contexts.

## Rationale: how does this CIP achieve its goals?

The proposed approach allows wallets to provide a complete and accurate summary of transactions without prohibiting transaction chaining or requiring dApps to change the way they interface with wallets.

Although it cannot entirely reduce the risk of a malicious actor tricking a user into signing a transaction that they didn't intend to, it will hopefully reduce the risk by showing an accurate summary of the transaction more often and, when not possible, displaying a clear warning to the user.

## Path to Active

### Acceptance Criteria

In order for this standard to be considered 'Active', the following criteria must be met:

- [ ] A thorough review by all relevant stakeholders
- [ ] Implementation in at least two Cardano wallets

### Implementation Plan

> N/A

Reviewer: What about demonstrating this in the extension namespace as a reference implementation?

Author: The plan was to make this a fundamental change to the web-wallet bridge standards themselves. Leaving CIP-30's `signTx` and CIP-103's `signTxs` endpoint as is would leave the attack vector available malicious users and render this CIP largely unhelpful.

I am happy to rewrite the CIP to clarify that. I tried to make it clear that, if a wallet implements this CIP, the wallet should implement transaction caching for _any_ current or future transaction signing functionality.

Reviewer: I wouldn't suggest rewriting the CIP... this question has just been for my own understanding 🙏

> ## Copyright
> 
> [CC-BY-4.0]: https://creativecommons.org/licenses/by/4.0/legalcode

Reviewer: ```suggestion
## Copyright

This CIP is licensed under [CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/legalcode).
```
